File: ex/hexagonal/src/main/kotlin/com/ktb/shop/application/RegisterCustomer.kt
package com.ktb.shop.application

import com.ktb.shop.domain.Customer
import com.ktb.shop.domain.CustomerId
import com.ktb.shop.domain.Email
import com.ktb.shop.domain.EmailAlreadyUsed
import com.ktb.shop.domain.port.CustomerRepository
import com.ktb.shop.domain.port.Notifier

data class RegisterCommand(val id: String, val email: String, val name: String)

// The use case — depends only on ports. No framework, no DB, no web.
class RegisterCustomer(
    private val customers: CustomerRepository,
    private val notifier: Notifier,
) {
    fun handle(command: RegisterCommand): Customer {
        val email = Email(command.email)
        if (customers.findByEmail(email) != null) throw EmailAlreadyUsed(command.email)

        val customer = Customer(CustomerId(command.id), email, command.name)
        customers.save(customer)
        notifier.welcome(customer)
        return customer
    }
}

File: ex/hexagonal-go/internal/application/register_customer.go
package application

import (
	"context"

	"github.com/ktb/hexagonal-shop/internal/domain"
	"github.com/ktb/hexagonal-shop/internal/domain/port"
)

type RegisterCommand struct {
	ID    string
	Email string
	Name  string
}

// RegisterCustomer is the use case — depends only on ports. No framework, no DB, no web.
type RegisterCustomer struct {
	customers port.CustomerRepository
	notifier  port.Notifier
}

func NewRegisterCustomer(customers port.CustomerRepository, notifier port.Notifier) *RegisterCustomer {
	return &RegisterCustomer{customers: customers, notifier: notifier}
}

func (uc *RegisterCustomer) Handle(ctx context.Context, cmd RegisterCommand) (domain.Customer, error) {
	email, err := domain.NewEmail(cmd.Email)
	if err != nil {
		return domain.Customer{}, err
	}

	existing, err := uc.customers.FindByEmail(ctx, email)
	if err != nil {
		return domain.Customer{}, err
	}
	if existing != nil {
		return domain.Customer{}, &domain.EmailAlreadyUsedError{Email: cmd.Email}
	}

	id, err := domain.NewCustomerID(cmd.ID)
	if err != nil {
		return domain.Customer{}, err
	}
	customer, err := domain.NewCustomer(id, email, cmd.Name)
	if err != nil {
		return domain.Customer{}, err
	}

	if err := uc.customers.Save(ctx, customer); err != nil {
		return domain.Customer{}, err
	}
	uc.notifier.Welcome(ctx, customer)
	return customer, nil
}
